tendermint/src/abci/data.rs

```diff
@@ -1,64 +1,78 @@
-use crate::{Error, Kind};
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
-use std::{
-    fmt::{self, Display},
-    str::FromStr,
-};
-use subtle_encoding::hex;
+use crate::Error;
+use serde::{Deserialize, Serialize};
+use std::convert::{TryFrom, TryInto};
+use tendermint_proto::serializers::bytes::Base64;
 
 /// ABCI transaction data.
 ///
 /// Transactions are opaque binary blobs which are validated according to
 /// application-specific rules.
-#[derive(Clone, Debug, Eq, PartialEq, Default)]
+#[derive(Clone, Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+#[serde(try_from = "Base64", into = "Base64")]
 pub struct Data(Vec<u8>);
 
-impl Data {
-    /// Borrow the data as bytes
-    pub fn as_bytes(&self) -> &[u8] {
```

Comment:
Is it perhaps not more readable/explicit to keep this method (`as_bytes()`) instead of implementing the `AsRef` trait?

```rust
// Option 1 (current):
let data_vec: Vec<u8> = data.as_ref().to_vec();

// Option 2 (proposed):
let data_vec: Vec<u8> = data.as_bytes().to_vec();
```

Reply:
I like using traits, especially if they are built into the language. In this case, both the `as_bytes()` function and the `AsRef` trait was implemented and most probably we are going to use neither.

I've removed both and added a getter (`value()`) instead. We've been adding constructors and getters lately, so this might be even easier to read.